core/client/fs/src/main/java/alluxio/client/block/stream/BufferCachingGrpcDataReader.java

```diff
@@ -0,0 +1,249 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.stream;
+
+import alluxio.client.file.FileSystemContext;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.grpc.ReadRequest;
+import alluxio.grpc.ReadResponse;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.network.protocol.databuffer.NioDataBuffer;
+import alluxio.resource.CloseableResource;
+import alluxio.wire.WorkerNetAddress;
+import alluxio.resource.LockResource;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A gRPC data reader that responsible for reading from a specific block.
+ *
+ * The current implementation cached the block data from the beginning to
+ * the largest index being read.
+ */
+@NotThreadSafe
+public class BufferCachingGrpcDataReader {
```

Comment:
It is strange to me that something called `BufferCachingGrpcDataReader` is not implementing `DataReader`. Is there a reason for that?

Also, is a unittest possible for this?

Reply:
I asked the same question in its PR and the author said it is very different from other `DataReader`, indicating it does not make sense to inherit `DataReader`.
Let's aim to add test in #12474 too